app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlInteractor.kt

```diff
@@ -322,6 +330,21 @@ class SessionControlInteractor(
         controller.handleReadPrivacyNoticeClicked()
     }
 
+    override fun showWallpapersOnboardingDialog(state: AppState?): Boolean {
+        return state?.wallpaperState?.availableWallpapers?.let { wallpapers ->
+            wallpapers.filter { wallpaper ->
+                wallpaper.thumbnailFileState == Wallpaper.ImageFileState.Downloaded
+            }.size.let { downloadedCount ->
+                downloadedCount >= WallpaperOnboardingDialogFragment.THUMBNAILS_COUNT &&
+                    downloadedCount == wallpapers.size
+            }.also { isThumbnailsDownloaded ->
+                if (isThumbnailsDownloaded) {
+                    controller.handleShowWallpapersOnboardingDialog()
+                }
+            }
```

Comment:
I think it would make more sense to move this logic into `SessionControlController`. This suggestion may help with readability as well

```suggestion
if (wallpapers.size >= WallpaperOnboardingDialogFragment.THUMBNAILS_COUNT && wallpapers.all { it.thumbnailFileState == Wallpaper.ImageFileState.Downloaded }) {
   controller.handleShowWallpapersOnboardingDialog()
}
```

Reply:
👍